src/did-manager/cheqd-did-provider.ts

```diff
@@ -104,16 +109,336 @@ export type TImportableEd25519Key = Required<Pick<IKey, 'publicKeyHex' | 'privat
+		for (const controller of controllers) {
+			// We need to get here current version of DIDDocument associated with the controller and cannot skip it even if document.id === controller
+			// cause in case of remooving verifcation method we need to sign the payload with the old verification method which is on ledger.
+			const controllerResolutionResult = await this.context.agent.resolveDid({ didUrl: controller });
+			const controllerDidDocument = controllerResolutionResult.didDocument;
+			// Check if controller DID document is resolved
+			if (!controllerDidDocument) {
+				throw new CheqdProviderError(
+					'[did-provider-cheqd]: compileSignInfos: Error while resolving the DID document for controller DID: ' + controller,
+					CheqdProviderErrorCodes.UnresolvedDID);
+			}
+			// Check that controller's DIDDocument is active
+			if (controllerResolutionResult.didDocumentMetadata.deactivated) {
+				throw new CheqdProviderError(
+					`[did-provider-cheqd]: compileSignInfos: DIDDocument associated with controller ${controller} is deactivated`,
+					CheqdProviderErrorCodes.DeactivatedController);
+			}
+			// Check if controller DID document contains verification methods
+			if (!controllerDidDocument.verificationMethod) {
+				throw new CheqdProviderError(
+					'[did-provider-cheqd]: compileSignInfos: Controller DID document does not contain verification methods',
+					CheqdProviderErrorCodes.EmptyVerificationMethod);
+			}
+			// Iterate over verificationMethods and by comparing publicKeys get the verificationMethod Id
+			for (const vm of controllerDidDocument.verificationMethod) {
+				// Try to match verificationMethod with one of publicKeyHexs
+				const verificationMethodId = ( function (){
+					for (const publicKeyHex of publicKeyHexs) {
+						// Transform to string
+						const publicKey = fromString(publicKeyHex, 'hex');
+						switch (vm?.type) {
+							case VerificationMethods.Ed255192020: {
+								const publicKeyMultibase = toMultibaseRaw(publicKey);
+								if (vm.publicKeyMultibase === publicKeyMultibase) {
+									return {
+										verificationMethodId: vm.id,
+										publicKeyHex: publicKeyHex
+									}
+								}
+								break;
+							}
+							case VerificationMethods.Ed255192018: {
+								const publicKeyBase58 = bases['base58btc'].encode(publicKey).slice(1);
+								if (vm.publicKeyBase58 === publicKeyBase58) {
+									return {
+										verificationMethodId: vm.id,
+										publicKeyHex: publicKeyHex
+									}
+								}
+								break;
+							}
+							case VerificationMethods.JWK: {
+								const publicKeyJwk: JsonWebKey = {
+									crv: 'Ed25519',
+									kty: 'OKP',
+									x: toString(publicKey, 'base64url'),
+								};
+								if (JSON.stringify(vm.publicKeyJwk) === JSON.stringify(publicKeyJwk)) {
+									return {
+										verificationMethodId: vm.id,
+										publicKeyHex: publicKeyHex
+									}
+								}
+								break;
+							}
+						}
+					}
+				})()
```

Comment:
Instead of doing explicit conversion from each verification method type, use the `extractPublicKeyHex` utility function exported by `@veramo/utils`, specifically meant for that purpose.

It'll reduce the overhead of direct dependencies + imports, since it handles the cryptographic primitives required internally.

That'll reduce the iteration closure to a few lines.